### PR TITLE
Allow comparison between different Iterable types.

### PR DESCRIPTION
**Pull request by technorama**
Fixed incorrect array comparison results.
Fixed Enumerable spec.
Added Iterable spec.
StaticArray is now Iterable.
== and equals? moved from Array to Iterable

### CODE REVIEW

File: src/iterable.cr [hunk 1]
```diff
@@ -33,4 +33,39 @@ module Iterable
   def each_with_object(obj)
     each.with_object(obj)
   end
+
+  # :nodoc:
+  # overrides Comparable.==
+  def ==(other: self)
+# TODO: if T is a noncontainer value && Enumerable is allocated as an array use memcmp()
```
Comment (jhass) [inline src/iterable.cr]: I don't think that'll work since we consider `1 == 1.0` to be true.

File: src/iterable.cr [hunk 2]
```diff
@@ -33,4 +33,39 @@ module Iterable
...
+    equals?(other) { |x, y| x == y }
+  end
+
+  # Iterates over both Enumerables comparing each value against other.value.
```
Comment (jhass) [inline src/iterable.cr]: There's no `other.value` method (called). Let's try to stay with explanatory wording here, not pseudo code.

File: src/array.cr [hunk 3]
```diff
@@ -852,6 +827,10 @@ class Array(T)
     @size = size.to_i
   end
 
+  def bytesize
+    sizeof(T) * @size
+  end
```
Comment (jhass) [inline src/array.cr]: What are the usecases for this method on `Array`? Shouldn't we encourage usage of Slice etc for these?

**Comment by technorama** [comment]
`bytesize` is added for future optimizing of comparison operations (using memcmp) which I will add when I can figure out how to reliably tell if the object can reference other objects.

I view Array/Slice as mostly interchangeable.  They already have a very high level of method compatibility.

**Comment by jhass** [comment]
But memcmp can only compare types of the same type, thus their size will equal if their bytesize equals.

**Comment by technorama** [comment]
1. Quick introspection (how much memory is this thing using?)
2. IO.read/write requires UInt8.  I don't see why that restriction is in place.  write(Slice(Struct Foo)) should also work.  `bytesize` can be used for reliable calculation.
3. Consistency.  Other methods have it (like String).

File: src/iterable.cr [hunk 4]
```diff
@@ -33,4 +33,40 @@ module Iterable
...
     each.with_object(obj)
   end
+
+  # :nodoc:
+  # Overrides Comparable.==
+  # This version works with more objects because it doesn't rely on <=>.
+  def ==(other: self)
+    # TODO: if T is a noncontainer value && Enumerable is allocated as an array use memcmp()
+    equals?(other) { |x, y| x == y }
+  end
```
Comment (jhass) [inline src/iterable.cr]: Perhaps we should fix `Comparable#==` or more generally `Object#<=>` instead. Ruby defines `Object#<=>` as returning `nil`:
> Your implementation of <=> should return one of the following values: -1, 0,
1 or nil. -1 means self is smaller than other. 0 means self is equal to other.
1 means self is bigger than other. Nil means the two values could not be
compared.

**Comment by technorama** [comment]
Or we can ditch implementing == everywhere, only require <=> and include Comparable in object.

Sharp Regrets: Worst C# feature `#9`.
http://www.informit.com/articles/article.aspx?p=2425867

**Comment by asterite** [comment]
I don't think different iterable objects should be always comparable. Array, Tuple and StaticArray between them: probably yes. Also don't forget that checking the length makes a big difference to quickly say false.

**Comment by asterite** [comment]
Equality and comparison are different things. Some things can be equal or different, but not necessarily comparables. For example ast nodes.

**Comment by technorama** [comment]
@asterite Is there an easy way to tell if the object has a fixed size and `size` is valid?  Is respond_to?(:size) enough?

Is there a way to tell if a type can contain references to other objects?

Both of those are needed for optimizing to memcmp instead of equality checks.

**Comment by technorama** [comment]
@asterite jhass is waiting to see if this patch is acceptable based on equality checking.

**Comment by asterite** [comment]
As I said, I don't think iterables should be comparable just like that. An Iterable can be infinite, for example, and this will hang the application.

I would make Array, Tuple and StaticArray comparables between themselves, because the standard library defines them and they are well known. I wanted to do this in 0.7.6 bug a bug prevented me to do so.

**Comment by asterite** [comment]
But let me check with @waj this week and I'll come back to you.

**Comment by waj** [comment]
Don't this makes every array comparison slower? Did you run some benchmarks. I think an algorithm to test equality between arbitrary enumerables should exist somewhere but I'm not sure about using it for arrays.

**Comment by technorama** [comment]
@asterite PartialComparable already handles the situation where 2 objects are not > or < but are also not equal.

Why not make <=> the single source of truth in Object and everything can be Comparable?  If there is an optimized `==` method the subclass can override it.  Most of the time there won't be.

**Comment by technorama** [comment]
@waj Iterable == can be optimized to a single memcmp if I can get the following information:
* How can I tell if a type contains references to other objects?  Specifically reference type objects.

A struct with ints should be safe to memcmp.
A struct with a single reference object would have to fall back to Iteration with == comparisons.

The old array comparison also iterated and used ==.  Why would this be significantly slower?

**Comment by technorama** [comment]
@asterite If the Iterable is infinite any operation that calls each will hang the application.  If this is truly a concern I can move the iterable comparison to a different module and only include the comparison operations in those 3 classes.

**Comment by waj** [comment]
It could be slower because using iterators sometimes introduce a small overhead, althought it might not be the case here. On the other hand, if the two arrays have different size there is no need to compare every element until it reaches the end of one of them. 

I think the condition to figure out if memcmp can be used or not depends on if the default comparison has been redefined. For example you can define a struct type with only value types as members, but still make different instances with different values, equal under the comparison.

**Comment by technorama** [comment]
@waj Updated to use memcmp only with Uint8 for binary data comparisons.

File: src/iterable.cr [hunk 5]
```diff
@@ -33,4 +33,54 @@ module Iterable
...
+    each do |item|
+      other_item = other_iter.next
+      return false if other_item.is_a?(Iterator::Stop)
+      return false unless yield(item, other_item)
+    end
+    return true if other_iter.next.is_a?(Iterator::Stop)
+    false
+  end
+
+  module FixedSizeCompare(T)
```
Comment (jhass) [inline src/iterable.cr]: If we really do want this it should be below `Comparable` IMO

File: src/iterable.cr [hunk 6]
```diff
@@ -33,4 +33,54 @@ module Iterable
...
+    def ==(other: self)
+      return false if size != other.size
+
+      # optimize binary data comparisons
+      if T.is_a?(UInt8.class)
+        bs = LibC::SizeT.new(bytesize)
+        LibC.memcmp(to_unsafe as Void*, other.to_unsafe as Void*, bs) == 0
+        equals?(other) { |x, y| x == y }
```
Comment (jhass) [inline src/iterable.cr]: I don't think you meant to include this line here.

File: src/iterable.cr [hunk 7]
```diff
@@ -33,4 +33,54 @@ module Iterable
...
+    return true if other_iter.next.is_a?(Iterator::Stop)
+    false
+  end
+
+  module FixedSizeCompare(T)
+    def ==(other: self)
+      return false if size != other.size
+
+      # optimize binary data comparisons
+      if T.is_a?(UInt8.class)
```
Comment (waj) [inline src/iterable.cr]: Having this special case for only this type just doesn't smell good to me

**Comment by technorama** [comment]
What would it take to get this PR through?
* Move the comparisons to enumerable?
* Make Enumerable.== check size?
* Other?


@waj Special optimizations exist in other classes:

string.cr
```
def ==(other : self)
  return true if same?(other)
  return false unless bytesize == other.bytesize

  cstr.memcmp(other.cstr, bytesize) == 0
end
```

Also enum.cr and tuple.cr.

**Comment by waj** [comment]
> What would it take to get this PR through?

I would not compare enumerables like this. The semantics of `==` is more related with equivalence of objects. Instead we could add a class method to Enumerable, similar to SequenceEqual in .NET: https://msdn.microsoft.com/en-us/library/vstudio/bb348567(v=vs.100).aspx

> Special optimizations exist in other classes

What I was saying is that with this change, because you deleted the `Array#==(Array)` method it will now use the generic implementation in Enumerable, and it will compare each element until it figures out the arrays have different size.

File: spec/std/enumerable_spec.cr [hunk 8]
```diff
@@ -262,7 +262,7 @@ describe "Enumerable" do
   describe "in_groups_of" do
     assert { [1, 2, 3].in_groups_of(1).should eq([[1], [2], [3]]) }
     assert { [1, 2, 3].in_groups_of(2).should eq([[1, 2], [3, nil]]) }
-    assert { ([] of Int32).in_groups_of(2).should eq([] of Array(Array(Int32 | Nil))) }
+    assert { ([] of Int32).in_groups_of(2).should eq([] of Array(Int32 | Nil)) }
```
Comment (waj) [inline spec/std/enumerable_spec.cr]: Why this change? `in_groups_of` should always return an array of arrays.

**Comment by waj** [comment]
Ok, my bad... I just found about the `FixedSizeCompare` module

File: src/iterable.cr (hunk 5 above)
Comment (technorama) [inline src/iterable.cr]: Comparable requires <=>.  Doing so would restrict the types of things that could be compared.

This uses == for greater compatibility with all objects.

File: spec/std/enumerable_spec.cr (hunk 8 above)
Comment (technorama) [inline spec/std/enumerable_spec.cr]: Not only is the test incorrect but the current in tree `==` is incorrect.  The current `==` identifies a 2 level deep empty array as equal to a 3 level deep empty array.  This PR fixes the broken behavior and corrects the test.

`[] of Array(Array(Int32 | Nil))` is a three level deep array.  `Array(Array(Array(Int32 | Nil)))`

Proof
```
a = [] of Array(Array(Int32 | Nil))
b = [] of Array(Int32 | Nil)
p a
p b
p a == b #=> says true but NOT true
```

`in_groups_of` works properly
```
# Both return Array(Array(Nil | Int32))
p ([1, 2, 3, 4] of Int32).in_groups_of(2) #=> [[1, 2], [3, 4]]
p ([] of Int32).in_groups_of(2)  #=> []
```

Comment (waj) [inline spec/std/enumerable_spec.cr]: Ok I can see the test is wrong. However I think the behaviour of comparison of arrays of different type was intentional.

Comment (asterite) [inline spec/std/enumerable_spec.cr]: I noticed the mistake when I accepted the pull request but forgot to fix it. Comparison between arrays like it is working now is OK.

**Comment by asterite** [comment]
@technorama Comparing iterable types with `each` is bad, you are creating two iterators (allocating memory) for this comparison.

I really like to do things one step at a time. We can start by making Array, Tuple and StaticArray comparable between themselves. Another idea is to have a `RandomAccess` module that you can include and that provides comparisons between RandomAccess types (they provide a `size` and `[]` methods). But trying to do everything so generically hurts in the long run.

**Comment by technorama** [comment]
Comparisons are now in Enumerable::FixedSizeCompare, not to be confused with Comparable which defines tons of compare operators all backed by <=>.

Array, StaticArray, & Slice all include FixedSizeCompare and may be tested for equality against each other.

**Comment by asterite** [comment]
@technorama The main issue is that not all iterables should be comparable between themselves. We aren't even sure an Array should equal StaticArray or Slice.

A more convincing example. Set is Iterable. In Ruby you do:

```ruby
require "set"

[1, 2, 3] == Set[1, 2, 3] #=> false
```

Ruby is right. In Crystal, accepting this pull request, it will probably say "true", but it's wrong because a Set has no order, even though iterating it has some order (order of insertion).

I'm closing this. We could have, if we want it, a method on Enumerable or Iterable to compare enumerables or iterables between them. But we first need a motivating use case.

**Comment by technorama** [comment]
@asterite I've constantly updated the code to take account of suggestions by others especially @jhass and @waj.  It doesn't compare any Iterable anymore.  It doesn't use Iterable at all.  There is a module included ONLY by Array, StaticArray & Slice `Enumerable::FixedSizeCompare` which handles comparisons within and between those 3.

The use cases are:
* Reducing code and complexity.  A single == is defined that works for all 3 classes.  Rather than error prone implementations in each one.
* Comparing specs correctly.  There was already an error caught by switching to the comparison methods in this PR.  This PR correctly compares Enumerables against each other (the current code base doesn't).
* Providing a uniform interface between 3 almost interchangeable classes.  `equals` only existed on Array, now it's in 1 module used by all 3 classes.
* Comparing bytes in an Array|StaticArray|Slice against Array|StaticArray|Slice
```
# MAGIC_HEADER could be declared in a number of ways
MAGIC_HEADER = Array(UInt8)new...
MAGIC_HEADER = Slice(UInt8).new...

# bytes could be declared in a number of ways for different purposes
bytes = [] of UInt8 # Used with read directly in a loop.
bytes = Array(Uint8).new # Used with read but data is passed around between fibers or other.
bytes = buffer[offset, len] # Slice of existing data from a larger packet or just declared as a Slice

# `bytes` and MAGIC_HEADER may cross library boundaries.  
# Developers on either side may change the type to optimize.
# When they do it will break on comparison at runtime (not compile).
raise "protocol error" if bytes != MAGIC_HEADER
```

**Comment by Nakilon** [comment]
> Ruby is right. In Crystal, accepting this pull request, it will probably say "true", but it's wrong because a Set has no order, even though iterating it has some order (order of insertion).

What about `Set#===`?
